core/trino-main/src/main/java/io/trino/operator/output/SlicePositionsAppender.java

```diff
@@ -184,25 +203,58 @@ private void copyBytes(Block block, int[] lengths, int[] positions, int count, i
     }
 
     /**
-     * Copy {@code length} bytes from {@code block}, at position {@code position} to {@code count} consecutive positions in the {@link #bytes} array.
+     * Copy {@code length} bytes from {@code rawSlice}, starting at offset {@code sourceOffset} to {@code count} consecutive positions in the {@link #bytes} array.
      */
-    private void duplicateBytes(Block block, int position, int count)
+    private void duplicateBytes(Slice rawSlice, int sourceOffset, int length, int count)
     {
-        int length = block.getSliceLength(position);
         int newByteCount = toIntExact((long) count * length);
         int startOffset = getCurrentOffset();
         ensureBytesCapacity(startOffset + newByteCount);
 
-        Slice slice = block.getSlice(position, 0, length);
+        if (rawSlice.hasByteArray()) {
+            byte[] base = rawSlice.byteArray();
+            int byteArrayOffset = rawSlice.byteArrayOffset();
+            for (int i = 0; i < count; i++) {
+                System.arraycopy(base, byteArrayOffset + sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
+        else {
+            for (int i = 0; i < count; i++) {
+                rawSlice.getBytes(sourceOffset, bytes, startOffset + i * length, length);
+            }
+        }
         for (int i = 0; i < count; i++) {
-            slice.getBytes(0, bytes, startOffset + (i * length), length);
             offsets[positionCount + i + 1] = startOffset + ((i + 1) * length);
         }
 
         positionCount += count;
         updateSize(count, newByteCount);
     }
 
+    private void appendGenericBlock(IntArrayList positions, Block block)
+    {
+        int newByteCount = 0;
+        for (int i = 0; i < positions.size(); i++) {
+            int position = positions.getInt(i);
+            if (block.isNull(position)) {
+                offsets[positionCount + 1] = offsets[positionCount];
+                valueIsNull[positionCount] = true;
+                hasNullValue = true;
+            }
+            else {
+                int length = block.getSliceLength(position);
+                ensureBytesCapacity(offsets[positionCount] + length);
```

Comment:
`offsets[positionCount]` -> `getCurrentOffset`.

but maybe `ensureBytesCapacity` should only be used with delta rather than absolute value, like `ensureExtraBytesCapacity`